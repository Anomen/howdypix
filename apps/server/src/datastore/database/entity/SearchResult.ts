import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
