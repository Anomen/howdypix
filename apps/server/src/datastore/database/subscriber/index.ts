export * from "./updateStoreOnPhotoSave";
export * from "./updateStoreOnAlbumSave";
export * from "./updateStoreOnSearchSave";
export * from "./updateStoreOnSearchResultSave";
