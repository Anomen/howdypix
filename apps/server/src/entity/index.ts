export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
