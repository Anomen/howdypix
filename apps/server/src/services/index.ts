export { startRabbitMq } from "./startRabbitMq";
export { startCacheDB } from "./startCacheDB";
export { startFileScan } from "./startFileScan";
