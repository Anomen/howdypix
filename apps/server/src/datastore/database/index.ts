import * as entities from "./entity";
import * as subscribers from "./subscriber";

export { entities, subscribers };
