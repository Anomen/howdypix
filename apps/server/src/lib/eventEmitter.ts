import StrictEventEmitter from "strict-event-emitter-types";
import { EventEmitter } from "events";
import { HFile, QueueData, QueueName } from "@howdypix/shared-types";
import { Photo } from "../datastore/database/entity";

export interface EventTypes {
  newDirectory: { root: string; hfile: HFile };
  unlinkDirectory: { root: string; hfile: HFile };
  newFile: { root: string; hfile: HFile };
  removeFile: { root: string; hfile: HFile };
  processFile: { root: string; hfile: HFile };
  processedFile: QueueData[QueueName.PROCESSED];

  newPhotoEntry: {
    data: Photo;
  };
}

export type Events = StrictEventEmitter<EventEmitter, EventTypes>;
