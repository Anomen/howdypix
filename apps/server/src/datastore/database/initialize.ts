import { Connection, EntitySchema, getConnectionManager } from "typeorm";
import { map } from "lodash";
import { AppStore } from "../state";
import * as subscribers from "./subscriber";
import * as entities from "./entity";
import { AppConfig } from "../../lib/config";
import { Events } from "../../lib/eventEmitter";

export const initializeDatabase = (
  dbConfig: AppConfig["database"],
  store: AppStore,
  event: Events
): Connection =>
  getConnectionManager().create({
    type: "sqlite",
    database: dbConfig.path,
    synchronize: true,
    logging: false,
    entities: map(entities, (entity) => (entity as unknown) as EntitySchema),
    subscribers: map(
      subscribers,
      (subscriber) => subscriber(event, store) as Function
    ),
  });
