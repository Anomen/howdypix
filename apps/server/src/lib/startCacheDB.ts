import { Connection } from "typeorm";
import { appDebug, generateThumbnailPaths, hjoin } from "@howdypix/utils";
import { existsSync, statSync, unlinkSync } from "fs";
import { join, parse } from "path";
import { UserConfig } from "../config";
import { Photo } from "../entity/Photo";
import { Events, EventTypes } from "./eventEmitter";
import { Source } from "../entity/Source";
import { Album } from "../entity/Album";

export async function onNewDir(
  { root, hfile }: EventTypes["newDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  const sourceRepository = connection.getRepository(Source);
  const album = await albumRepository.findOne({ where: { inode: stat.ino } });

  // Check the updated_date
  if (!album) {
    const newAlbum = new Album();
    newAlbum.dir = hfile.file ?? "";
    newAlbum.parentDir = hfile.dir ?? "";
    newAlbum.source = hfile.source;
    newAlbum.inode = stat.ino.toString();

    newAlbum.sourceLk =
      (await sourceRepository.findOne({
        source: hfile.source,
      })) ?? null;

    albumRepository.save(newAlbum);
  }
}

export async function onUnlinkDir(
  { root, hfile }: EventTypes["unlinkDirectory"],
  event: Events,
  connection: Connection
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const albumRepository = connection.getRepository(Album);
  await albumRepository.delete({ inode: stat.ino.toString() });
}

export async function onNewFile(
  { root, hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const absolutePath = join(root, hfile.dir ?? "", hfile.file ?? "");
  const stat = statSync(absolutePath);

  // Check if it exists in the database
  const photoRepository = connection.getRepository(Photo);
  const photo = await photoRepository.findOne({ inode: stat.ino });

  // Check the updated_date
  if (photo?.mtime === stat.mtimeMs) {
    const thumbnailsExist: boolean = generateThumbnailPaths(
      userConfig.thumbnailsDir,
      hfile
    ).reduce(
      (accumulator: boolean, tn) => accumulator && existsSync(tn.path),
      true
    );

    if (thumbnailsExist) {
      return;
    }
  }

  // Act
  event.emit("processFile", { root, hfile });
}

export async function onRemoveFile(
  { hfile }: EventTypes["newFile"],
  event: Events,
  connection: Connection,
  userConfig: UserConfig
): Promise<void> {
  const photoRepository = connection.getRepository(Photo);

  // Remove from database
  await photoRepository.delete(hfile);

  // Remove thumbnails
  generateThumbnailPaths(userConfig.thumbnailsDir, hfile).forEach((tn) => {
    unlinkSync(tn.path);
  });
}

export async function onProcessedFile(
  file: EventTypes["processedFile"],
  event: Events,
  connection: Connection
): Promise<void> {
  const albumRepository = connection.getRepository(Album);
  const photoRepository = connection.getRepository(Photo);

  const photo = new Photo();
  photo.make = file.exif.make ?? "";
  photo.model = file.exif.model ?? "";
  photo.ISO = file.exif.ISO ?? 0;
  photo.shutter = file.exif.shutter ?? 0;
  photo.processedShutter = file.exif.shutter
    ? Math.round((1 / file.exif.shutter) * 10) / 10
    : 0;
  photo.aperture = file.exif.aperture ?? 0;
  photo.processedAperture = file.exif.aperture
    ? Math.round(file.exif.aperture * 10) / 10
    : 0;
  photo.createDate = file.exif.createDate ?? 0;
  photo.inode = file.stat.inode;
  photo.mtime = file.stat.mtime;
  photo.ctime = file.stat.ctime;
  photo.birthtime = file.stat.birthtime;
  photo.size = file.stat.size;
  photo.source = file.hfile.source;
  photo.parentDir = parse(file.hfile.dir ?? "").dir ?? "";
  photo.dir = file.hfile.dir ?? "";
  photo.file = file.hfile.file ?? "";

  photo.album =
    (await albumRepository.findOne({
      dir: file.hfile.dir ?? "",
      source: file.hfile.source,
      parentDir: parse(file.hfile.dir ?? "").dir ?? "",
    })) ?? null;

  await photoRepository.save(photo);
  appDebug("db")(`Saved ${hjoin(file.hfile)}.`);
}

export async function startCacheDB(
  event: Events,
  userConfig: UserConfig,
  connection: Connection
): Promise<Connection> {
  // Insert the source entries in the database
  await Promise.all(
    Object.keys(userConfig.photoDirs).map(async (source) => {
      await Source.upsert(source, userConfig.photoDirs[source]);
    })
  );

  event.on("newFile", (params) =>
    onNewFile(params, event, connection, userConfig)
  );

  event.on("removeFile", (params) =>
    onRemoveFile(params, event, connection, userConfig)
  );

  event.on("newDirectory", (params) => onNewDir(params, event, connection));

  event.on("unlinkDirectory", (params) =>
    onUnlinkDir(params, event, connection)
  );

  event.on("processedFile", (file) => onProcessedFile(file, event, connection));

  return connection;
}
