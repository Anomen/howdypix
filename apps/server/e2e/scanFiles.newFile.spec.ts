import { statSync } from "fs";
import { appError } from "@howdypix/utils";
import waitForExpect from "wait-for-expect";
import { startCacheDB } from "../src/services/startCacheDB";
import { Album, Photo } from "../src/datastore/database/entity";
import { Events, EventTypes } from "../src/lib/eventEmitter";
import { initialize } from "./scanFiles.setup";

jest.mock("@howdypix/utils", () => ({
  ...jest.requireActual("@howdypix/utils"),
  appError: jest.fn(),
}));

jest.mock("fs", () => ({
  ...jest.requireActual("fs"),
  statSync: jest.fn(),
}));

describe("[NewFile] Scanning files", (): void => {
  let events: Events;

  const {
    pathHashCode,
    baseUserConfig,
    retrievePhotos,
    retrieveAlbums,
    resetScanFilesTests,
    connection,
  } = initialize();

  const executeTest = async (
    eventParams: EventTypes["newFile"],
    expectedAlbumLen: number
  ): Promise<void> => {
    await Promise.all([
      // SHOULD SEND AN EVENT
      new Promise((resolve, reject): void => {
        events.once("processFile", (params): void => {
          try {
            expect(params).toMatchSnapshot();
            resolve();
          } catch (e) {
            reject(e);
          }
        });

        events.emit("newFile", eventParams);
      }),
      // SHOULD SAVE IN THE DATABASE
      await waitForExpect(
        async (): Promise<void> => {
          const photos = await retrievePhotos();
          const albums = await retrieveAlbums();

          expect(photos.length).toEqual(1);
          expect(photos).toMatchSnapshot();

          expect(albums.length).toEqual(expectedAlbumLen);
          expect(albums).toMatchSnapshot();
        }
      ),
    ]);
  };

  // ===================================================
  // Reset after each tests
  // ===================================================
  beforeEach(
    async (): Promise<void> => {
      jest.resetAllMocks();
      jest.restoreAllMocks();

      const setup = await resetScanFilesTests();
      events = setup.events;

      jest.spyOn(connection.getRepository(Photo), "save");
      jest.spyOn(Album, "insertIfDoesntExist");
      jest.spyOn(events, "emit");

      (statSync as jest.Mock).mockImplementation((name) => ({
        ino: pathHashCode(name),
        mtime: new Date(1),
        ctime: new Date(2),
        birthtime: new Date(3),
        size: 10,
      }));

      await startCacheDB(events, baseUserConfig, connection);
    }
  );

  test("should save a new entry and emit an event - album file", async (): Promise<
    void
  > =>
    executeTest(
      {
        root: "/home/lelouch/first",
        hfile: {
          source: "first",
          file: "PHOTO1.jpg",
          dir: "myAlbum",
        },
      },
      3
    ));

  test("should save a new entry and emit an event - sub-album file", async (): Promise<
    void
  > =>
    executeTest(
      {
        root: "/home/lelouch/first",
        hfile: {
          source: "first",
          file: "PHOTO1.jpg",
          dir: "myAlbum/mySubAlbum",
        },
      },
      3
    ));

  test("should save a new entry and emit an event - root file", async () =>
    executeTest(
      {
        root: "/home/lelouch/first",
        hfile: {
          source: "first",
          file: "PHOTO1.jpg",
          dir: "",
        },
      },
      2
    ));

  test("should not save anything if we cannot have the album information to link", async (): Promise<
    void
  > => {
    jest
      .spyOn(Album, "insertIfDoesntExist")
      .mockImplementation(async (): Promise<null> => null);

    // jest.spyOn(events, "emit");

    events.emit("newFile", {
      root: "/home/lelouch/first",
      hfile: {
        source: "first",
        file: "PHOTO1.jpg",
        dir: "",
      },
    });

    // SHOULD NOT SAVE IN THE DATABASE
    await waitForExpect((): void => {
      expect(appError).toHaveBeenCalled();
    });

    expect(await retrievePhotos()).toHaveLength(0);
    expect(await retrieveAlbums()).toHaveLength(2);
    expect(events.emit).toHaveBeenCalledTimes(1);
  });

  test("should only send event if the photo already exists in the DB", async (): Promise<
    void
  > => {
    const params: EventTypes["newFile"] = {
      root: "/home/lelouch/first",
      hfile: {
        source: "first",
        file: "PHOTO1.jpg",
        dir: "",
      },
    };

    await new Promise((resolve, reject): void => {
      events.once("processFile", (): void => {
        try {
          expect(appError).not.toHaveBeenCalled();

          (connection.getRepository(Photo).save as jest.Mock).mockClear();

          events.once("processFile", (): void => {
            try {
              expect(appError).not.toHaveBeenCalled();
              expect(
                connection.getRepository(Photo).save
              ).not.toHaveBeenCalled();
              resolve();
            } catch (e) {
              reject(e);
            }
          });

          // Emit a second time with the same file => should not save anything in the DB
          events.emit("newFile", params);
        } catch (e) {
          reject(e);
        }
      });

      events.emit("newFile", params);
    });
  });

  test("should fail gracefully", async (): Promise<void> => {
    (statSync as jest.Mock).mockImplementation((): void => {
      throw new Error("ENOT");
    });

    // jest.spyOn(events, "emit");

    events.emit("newFile", {
      root: "/home/lelouch/first",
      hfile: {
        source: "first",
        file: "PHOTO1.jpg",
        dir: "",
      },
    });

    await waitForExpect((): void => {
      expect(appError).toHaveBeenCalled();
    });

    expect(events.emit).toHaveBeenCalledTimes(1);
  });
});
