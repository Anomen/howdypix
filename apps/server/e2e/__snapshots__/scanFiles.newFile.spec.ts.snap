// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[NewFile] Scanning files should save a new entry and emit an event - album file 1`] = `
Object {
  "hfile": Object {
    "dir": "myAlbum",
    "file": "PHOTO1.jpg",
    "source": "first",
  },
  "root": "/home/lelouch/first",
}
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - album file 2`] = `
Array [
  Object {
    "ISO": null,
    "album": Object {
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "myAlbum",
      "id": 1,
      "inode": "Zmlyc3RteUFsYnVt",
      "parentDir": "",
      "photos": undefined,
      "source": "first",
      "sourceLk": undefined,
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "albumId": 3,
    "aperture": null,
    "birthtime": 0,
    "createDate": null,
    "createdAt": 2016-01-01T00:00:00.000Z,
    "ctime": 0,
    "dir": "myAlbum",
    "file": "PHOTO1.jpg",
    "id": 1,
    "inode": "bXlBbGJ1bVBIT1RPMS5qcGc=",
    "make": null,
    "model": null,
    "mtime": 0,
    "parentDir": "",
    "processedAperture": null,
    "processedShutter": null,
    "searchResults": undefined,
    "shutter": null,
    "size": 10,
    "source": "first",
    "status": "not_processed",
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
]
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - album file 3`] = `
Array [
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "myAlbum",
    "id": 1,
    "inode": "Zmlyc3RteUFsYnVt",
    "parentDir": "",
    "photos": Array [
      Object {
        "ISO": null,
        "album": undefined,
        "albumId": 3,
        "aperture": null,
        "birthtime": 0,
        "createDate": null,
        "createdAt": 2016-01-01T00:00:00.000Z,
        "ctime": 0,
        "dir": "myAlbum",
        "file": "PHOTO1.jpg",
        "id": 1,
        "inode": "bXlBbGJ1bVBIT1RPMS5qcGc=",
        "make": null,
        "model": null,
        "mtime": 0,
        "parentDir": "",
        "processedAperture": null,
        "processedShutter": null,
        "searchResults": undefined,
        "shutter": null,
        "size": 10,
        "source": "first",
        "status": "not_processed",
        "updatedAt": 2016-01-01T00:00:00.000Z,
      },
    ],
    "source": "first",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./first",
      "id": 1,
      "source": "first",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyZmlyc3Q=",
    "parentDir": "",
    "photos": Array [],
    "source": "first",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./first",
      "id": 1,
      "source": "first",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyc2Vjb25k",
    "parentDir": "",
    "photos": Array [],
    "source": "second",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./second",
      "id": 1,
      "source": "second",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
]
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - root file 1`] = `
Object {
  "hfile": Object {
    "dir": "",
    "file": "PHOTO1.jpg",
    "source": "first",
  },
  "root": "/home/lelouch/first",
}
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - root file 2`] = `
Array [
  Object {
    "ISO": null,
    "album": Object {
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "",
      "id": 1,
      "inode": "c2VydmVyZmlyc3Q=",
      "parentDir": "",
      "photos": undefined,
      "source": "first",
      "sourceLk": undefined,
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "albumId": 2,
    "aperture": null,
    "birthtime": 0,
    "createDate": null,
    "createdAt": 2016-01-01T00:00:00.000Z,
    "ctime": 0,
    "dir": "",
    "file": "PHOTO1.jpg",
    "id": 1,
    "inode": "Zmlyc3RQSE9UTzEuanBn",
    "make": null,
    "model": null,
    "mtime": 0,
    "parentDir": "",
    "processedAperture": null,
    "processedShutter": null,
    "searchResults": undefined,
    "shutter": null,
    "size": 10,
    "source": "first",
    "status": "not_processed",
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
]
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - root file 3`] = `
Array [
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyZmlyc3Q=",
    "parentDir": "",
    "photos": Array [
      Object {
        "ISO": null,
        "album": undefined,
        "albumId": 2,
        "aperture": null,
        "birthtime": 0,
        "createDate": null,
        "createdAt": 2016-01-01T00:00:00.000Z,
        "ctime": 0,
        "dir": "",
        "file": "PHOTO1.jpg",
        "id": 1,
        "inode": "Zmlyc3RQSE9UTzEuanBn",
        "make": null,
        "model": null,
        "mtime": 0,
        "parentDir": "",
        "processedAperture": null,
        "processedShutter": null,
        "searchResults": undefined,
        "shutter": null,
        "size": 10,
        "source": "first",
        "status": "not_processed",
        "updatedAt": 2016-01-01T00:00:00.000Z,
      },
    ],
    "source": "first",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./first",
      "id": 1,
      "source": "first",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyc2Vjb25k",
    "parentDir": "",
    "photos": Array [],
    "source": "second",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./second",
      "id": 1,
      "source": "second",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
]
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - sub-album file 1`] = `
Object {
  "hfile": Object {
    "dir": "myAlbum/mySubAlbum",
    "file": "PHOTO1.jpg",
    "source": "first",
  },
  "root": "/home/lelouch/first",
}
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - sub-album file 2`] = `
Array [
  Object {
    "ISO": null,
    "album": Object {
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "myAlbum/mySubAlbum",
      "id": 1,
      "inode": "bXlBbGJ1bW15U3ViQWxidW0=",
      "parentDir": "myAlbum",
      "photos": undefined,
      "source": "first",
      "sourceLk": undefined,
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "albumId": 3,
    "aperture": null,
    "birthtime": 0,
    "createDate": null,
    "createdAt": 2016-01-01T00:00:00.000Z,
    "ctime": 0,
    "dir": "myAlbum/mySubAlbum",
    "file": "PHOTO1.jpg",
    "id": 1,
    "inode": "bXlTdWJBbGJ1bVBIT1RPMS5qcGc=",
    "make": null,
    "model": null,
    "mtime": 0,
    "parentDir": "myAlbum",
    "processedAperture": null,
    "processedShutter": null,
    "searchResults": undefined,
    "shutter": null,
    "size": 10,
    "source": "first",
    "status": "not_processed",
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
]
`;

exports[`[NewFile] Scanning files should save a new entry and emit an event - sub-album file 3`] = `
Array [
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "myAlbum/mySubAlbum",
    "id": 1,
    "inode": "bXlBbGJ1bW15U3ViQWxidW0=",
    "parentDir": "myAlbum",
    "photos": Array [
      Object {
        "ISO": null,
        "album": undefined,
        "albumId": 3,
        "aperture": null,
        "birthtime": 0,
        "createDate": null,
        "createdAt": 2016-01-01T00:00:00.000Z,
        "ctime": 0,
        "dir": "myAlbum/mySubAlbum",
        "file": "PHOTO1.jpg",
        "id": 1,
        "inode": "bXlTdWJBbGJ1bVBIT1RPMS5qcGc=",
        "make": null,
        "model": null,
        "mtime": 0,
        "parentDir": "myAlbum",
        "processedAperture": null,
        "processedShutter": null,
        "searchResults": undefined,
        "shutter": null,
        "size": 10,
        "source": "first",
        "status": "not_processed",
        "updatedAt": 2016-01-01T00:00:00.000Z,
      },
    ],
    "source": "first",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./first",
      "id": 1,
      "source": "first",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyZmlyc3Q=",
    "parentDir": "",
    "photos": Array [],
    "source": "first",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./first",
      "id": 1,
      "source": "first",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
  Object {
    "createdAt": 2016-01-01T00:00:00.000Z,
    "dir": "",
    "id": 1,
    "inode": "c2VydmVyc2Vjb25k",
    "parentDir": "",
    "photos": Array [],
    "source": "second",
    "sourceLk": Object {
      "albums": undefined,
      "createdAt": 2016-01-01T00:00:00.000Z,
      "dir": "./second",
      "id": 1,
      "source": "second",
      "updatedAt": 2016-01-01T00:00:00.000Z,
    },
    "updatedAt": 2016-01-01T00:00:00.000Z,
  },
]
`;
